contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java

```diff
@@ -0,0 +1,806 @@
+     * in {@link UnboundedReader#getWatermark()}.
+     *
+     * @param watermarkFn to calculate watermark at a record.
+     * @return Builder
+     */
+    public Builder<K, V> withWatermarkFn(
+        SerializableFunction<KafkaRecord<K, V>, Instant> watermarkFn) {
+      this.watermarkFn = Optional.of(watermarkFn);
+      return this;
+    }
+
+    /**
+     * Build Unbounded Kafka Source
+     *
+     * @return UnboundedKafkaSource
+     */
+    public UnboundedSource<KafkaRecord<K, V>, KafkaCheckpointMark> build() {
+
+      ImmutableMap<String, Object> consumerConfig = ImmutableMap.copyOf(mutableConsumerConfig);
+
+      Preconditions.checkNotNull(
+          consumerConfig.get(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG),
+          "Kafka bootstrap servers should be set");
+      Preconditions.checkNotNull(topics, "Kafka topics should be set");
+      Preconditions.checkArgument(!topics.isEmpty(), "At least one topic is required");
+      Preconditions.checkNotNull(keyCoder, "Coder for Kafka key bytes is required");
+      Preconditions.checkNotNull(valueCoder, "Coder for Kafka values bytes is required");
+
+      return new UnboundedKafkaSource<K, V>(
+          -1,
+          consumerConfig,
+          topics,
+          keyCoder,
+          valueCoder,
+          timestampFn,
+          watermarkFn,
+          ImmutableList.<TopicPartition>of() // no assigned partitions yet.
+          );
+    }
+  }
+
+  /** Static class, prevent instantiation. */
+  private KafkaSource() {}
+
+  private static class UnboundedKafkaSource<K, V>
+      extends UnboundedSource<KafkaRecord<K, V>, KafkaCheckpointMark> {
+
+    private final int id; // split id, mainly for debugging
+    private final ImmutableMap<String, Object> consumerConfig;
+    private final List<String> topics;
+    private final Coder<K> keyCoder;
+    private final Coder<V> valueCoder;
+    private final SerializableFunction<KafkaRecord<K, V>, Instant> timestampFn;
+    // would it be a good idea to pass currentTimestamp to watermarkFn?
+    private final Optional<SerializableFunction<KafkaRecord<K, V>, Instant>> watermarkFn;
+    private final List<TopicPartition> assignedPartitions;
+
+    public UnboundedKafkaSource(
+        int id,
+        ImmutableMap<String, Object> consumerConfig,
+        List<String> topics,
+        Coder<K> keyCoder,
+        Coder<V> valueCoder,
+        SerializableFunction<KafkaRecord<K, V>, Instant> timestampFn,
+        Optional<SerializableFunction<KafkaRecord<K, V>, Instant>> watermarkFn,
+        List<TopicPartition> assignedPartitions) {
+
+      this.id = id;
+      this.consumerConfig = consumerConfig;
+      this.topics = topics;
+      this.keyCoder = keyCoder;
+      this.valueCoder = valueCoder;
+      this.timestampFn = timestampFn;
+      this.watermarkFn = watermarkFn;
+      this.assignedPartitions = ImmutableList.copyOf(assignedPartitions);
+    }
+
+    @Override
+    public List<UnboundedKafkaSource<K, V>> generateInitialSplits(
```

Comment:
probably worth javadoccing the splitting behavior extensively here, and briefly at the class level (with a `{@link }` pointer to here). Users will want to understand how the source is scalabale